### ISSUE TITLE
Error while adding Gitlab external source 

### ISSUE THREAD

**Issue by sk-aaludra** [sourcegraph/sourcegraph#3526]
- **Sourcegraph version:** 3.3
- **Platform information:** CentOS 7, NTT Cloud, Chrome, Docker 18

#### Steps to reproduce:

Adding external source and while saving getting the below error

Error saving invalid configuration:

warning: external service updated, but sync request failed: syncer.sync.sourced: 1 error occurred:
* error listing GitLab projects: url="projects?membership=true&order_by=last_activity_at&per_page=100": Get https://gitlab.aaludradevelopers.com/api/v4/projects?membership=true&order_by=last_activity_at&per_page=100: dial tcp 180.12.181.9:443: i/o timeout

I'm getting results while executing the below passing token

curl -H 'Private-Token: <tokenstring>' -XGET 'https://gitlab.aaludradevelopers.com/api/v4/projects'

Can some one help?

**Comment by sk-aaludra** [comment]
After restarting the server the error is disappear, however getting error. There is nothing running on port 3182

Error saving invalid configuration:

warning: external service updated, but sync request failed: Post http://127.0.0.1:3182/sync-external-service: dial tcp 127.0.0.1:3182: connect: connection refused

**Comment by tsenart** [comment]
> warning: external service updated, but sync request failed: Post http://127.0.0.1:3182/sync-external-service: dial tcp 127.0.0.1:3182: connect: connection refused

That means that the restart of the service didn't succeed entirely. Can you find the logs respective to `repo-updater`? It should be listening on that port, but it seems like it didn't come up.

Additionally, how did you restart the server?

**Comment by sk-aaludra** [comment]
10:27:45       repo-updater | t=2019-04-22T10:27:45+0000 lvl=eror msg=store.done error="migrate.repos-enabled-state-deprecation.sources.list-repos: 1 error occurred:\n\t* error listing GitLab projects: url=\"projects?membership=true&order_by=last_activity_at&per_page=100\": Get https://gitlab.aaludradevelopers.com/api/v4/projects?membership=true&order_by=last_activity_at&per_page=100: dial tcp 180.12.181.9:443: i/o timeout\n\n"

I tried restarting the docker container, also from sourcegraph web interface

**Comment by tsenart** [comment]
As you can see by the error message, we're seeing i/o timeouts. Can you try running the exact same request with curl from **within** the sourcegraph Docker container?

```
docker exec -it $SOURCEGRAPH_CONTAINER_NAME /bin/bash
curl -v -H 'Private-Token: $TOKEN' -X GET 'https://gitlab.aaludradevelopers.com/api/v4/projects?membership=true&order_by=last_activity_at&per_page=100'
```

**Comment by sk-aaludra** [comment]
I have tried, from container its not able to reach the gitlab host, so far not sure why.

OK: 66 MiB in 55 packages
bash-4.4# curl -v -H 'Private-Token: <token>' -X GET 'https://gitlab.aaludradevelopers.com/api/v4/projects?membership=true&order_by=last_activity_at&per_page=100'
Note: Unnecessary use of -X or --request, GET is already inferred.
*   Trying 180.12.181.9...
* TCP_NODELAY set
^C
bash-4.4# ping google.com
PING google.com (172.217.161.206): 56 data bytes
64 bytes from 172.217.161.206: seq=0 ttl=50 time=14.021 ms
64 bytes from 172.217.161.206: seq=1 ttl=50 time=11.532 ms
^C
--- google.com ping statistics ---
3 packets transmitted, 3 packets received, 0% packet loss
round-trip min/avg/max = 11.532/13.879/16.084 ms
bash-4.4# ping 180.12.181.9
PING 180.12.181.9 (180.12.181.9): 56 data bytes
^C
--- 180.12.181.9 ping statistics ---
5 packets transmitted, 0 packets received, 100% packet loss

**Comment by tsenart** [comment]
Seems like you need to do some network debugging. What does a `traceroute` to that IP show you?

**Comment by sk-aaludra** [comment]
Apparently this is not an issue related to sourcegraph. I shall try to get in touch with cloud team to resolve the issue. Thanks for your prompt support so far

bash-4.4# traceroute gitlab.aaludradevelopers.com
traceroute to gitlab.aaludradevelopers.com (180.12.181.9), 30 hops max, 46 byte packets
 1  172.17.0.1 (172.17.0.1)  0.024 ms  0.026 ms  0.015 ms
 2  *  *  *
 3  *  *  *
 4  *  *  *
 5  *  *  *
 6  *  *  *
 7  *  *  *

traceroute to google.com (172.217.161.206), 30 hops max, 46 byte packets
 1  172.17.0.1 (172.17.0.1)  0.027 ms  0.020 ms  0.015 ms
 2  10.1.0.1 (10.1.0.1)  2.155 ms  1.757 ms  2.006 ms
 3  180-12-181-2-revip-jp5-default.181.12.180.in-addr.arpa (180.12.181.2)  2.951 ms  3.717 ms  2.982 ms
 4  61.199.130.81 (61.199.130.81)  3.330 ms  26.658 ms  13.045 ms
 5  122.1.245.65 (122.1.245.65)  4.114 ms  60.37.54.161 (60.37.54.161)  4.820 ms  23.276 ms
 6  60.37.54.166 (60.37.54.166)  4.897 ms  4.195 ms  4.080 ms

**Comment by slimsag** [comment]
Hi @sk-aaludra I am curious about some more details if you don't mind:

1. Could you let me know if you upgraded this instance from an older version of Sourcegraph or not?
2. Do your server logs show any indication of rate limiting? e.g. do you see anything from the following commands:

```
docker logs CONTAINER_ID 2>&1 | grep 'rate limit'
```

```
docker logs CONTAINER_ID 2>&1 | grep 'abuse detection'
```

**Comment by sk-aaludra** [comment]
Of course with pleasure -

This is a fresh installation, I found this interesting project from Digital Ocean and thought of giving a try. 

Logs doesn't shows any issues, find below for more details

CONTAINER ID        IMAGE                      COMMAND                  CREATED             STATUS              PORTS                                            NAMES
ba53f165ba50        sourcegraph/server:3.3.0   "/sbin/tini -- /usr/…"   4 days ago          Up 2 days           0.0.0.0:2633->2633/tcp, 0.0.0.0:7080->7080/tcp   pensive_mayer
[root@aaludradevelopers ~]# docker logs ba53f165ba50  2>&1 | grep 'rate limit'
[root@aaludradevelopers ~]# docker logs ba53f165ba50 2>&1 | grep 'abuse detection'
[root@aaludradevelopers ~]#

Regarding the network issue, I could understand the problem, however found no solution yet.

Gitlab and sourcegraph in the same machine in NTT cloud. Assigned public IP in Juniper vSRX firewall and Gitlab is NAT'd to LAN IP. From Docker, some reason is not able to reach the host, I have tried adding host entry for gitlab.aaludradevelopers.com to LAN IP, still no luck.

**Comment by sk-aaludra** [comment]
Network issues is resolved. Within Docker container, I'm able to pull the data from Gitlab, however sourcegraph web console, throwing the same error. PFA, any help?

[sourcegraph.pdf](https://github.com/sourcegraph/sourcegraph/files/3119842/sourcegraph.pdf)

**Comment by tsenart** [comment]
What is this attachment?

**Comment by sk-aaludra** [comment]
Sourcegraph Error details, and container log. FYR

**Comment by tsenart** [comment]
Please share it in a Github gist.

**Comment by sk-aaludra** [comment]
Here you go https://gist.github.com/sk-aaludra/ca1982b7b2bfc534b967e77538c24579

**Comment by tsenart** [comment]
This seems to be the same issue still. How was the network fixed?

**Comment by sk-aaludra** [comment]
Added host entry for gitlab domain to LAN IP address

**Comment by tsenart** [comment]
I opened a PR: https://github.com/sourcegraph/sourcegraph/pull/3632

**Comment by tsenart** [comment]
@sk-aaludra: Tomorrow or Monday, please try running the `sourcegraph/server:insiders` image and let us know if it fixed your issue.

**Comment by sk-aaludra** [comment]
Awesome, it works as expected. Let me start exploring this interesting project. Thanks for the quick turnaround. I hope this fix will be included in future release

**Comment by tsenart** [comment]
It will be included in the 3.4 release. Cheers!
